File: route-persistence/src/main/java/com/skedgo/routepersistence/RouteRepository.java
package com.skedgo.routepersistence;

import android.support.annotation.NonNull;
import android.support.v4.util.LruCache;

import com.skedgo.android.common.model.TripGroup;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

import hugo.weaving.DebugLog;
import rx.Observable;
import rx.functions.Func0;
import rx.subjects.PublishSubject;

public class RouteRepository {
  private final LruCache<String, CopyOnWriteArrayList<TripGroup>> memCache = new LruCache<>(1);
  private final PublishSubject<String> onNewTripGroupsAvailable = PublishSubject.create();
  private final RouteStore routeStore;
  private final Func0<Long> currentMillisProvider;

  public RouteRepository(
      RouteStore routeStore,
      Func0<Long> currentMillisProvider) {
    this.routeStore = routeStore;
    this.currentMillisProvider = currentMillisProvider;
  }

  @DebugLog public synchronized List<TripGroup> getTripGroups(String requestId) {
    final List<TripGroup> mem = memCache.get(requestId);
    if (mem != null) {
      return new ArrayList<>(mem);
    }

    final List<TripGroup> disk = routeStore.getTripGroupsByRouteIdAsync(requestId)
        .toList()
        .toBlocking()
        .first();
    memCache.put(requestId, new CopyOnWriteArrayList<>(disk));
    return disk;
  }

  public Observable<Integer> deletePastRoutesAsync() {
    return routeStore.deleteAsync(WhereClauses.happenedBefore(
        3, /* hours */
        currentMillisProvider.call()
    ));
  }

  @DebugLog public synchronized void addTripGroups(
      String requestId,
      @NonNull List<TripGroup> groups) {

    CopyOnWriteArrayList<TripGroup> existingGroups;
    existingGroups = memCache.get(requestId);
    if (existingGroups == null) {
      existingGroups = new CopyOnWriteArrayList<>();
      memCache.put(requestId, existingGroups);
    }

    existingGroups.addAll(groups);
    onNewTripGroupsAvailable.onNext(requestId);
  }

  public Observable<String> onNewTripGroupsAvailable() {
    return onNewTripGroupsAvailable.asObservable();
  }
}

File: route-persistence/src/main/java/com/skedgo/routepersistence/RouteRepository.kt
package com.skedgo.routepersistence

import rx.Observable
import rx.functions.Func0
import rx.subjects.PublishSubject
import skedgo.tripkit.routing.TripGroup
import java.util.concurrent.ConcurrentHashMap

open class RouteRepository constructor(
    private val routeStore: RouteStore,
    private val currentMillisProvider: Func0<Long>) {

  internal var requestIdTripGroupIdCache: ConcurrentHashMap<String, MutableList<String>> = ConcurrentHashMap()
  internal var tripGroupCache: ConcurrentHashMap<String, TripGroup> = ConcurrentHashMap()
  private val onNewTripGroupsAvailable = PublishSubject.create<String>()

  open fun getTripGroups(requestId: String): List<TripGroup> =
      if (requestIdTripGroupIdCache.containsKey(requestId)) {
        requestIdTripGroupIdCache[requestId]
            ?.filter { tripGroupCache.containsKey(it) }
            ?.map { tripGroupCache[it]!! }!!
      } else {
        routeStore.queryAsync(GroupQueries.hasRequestId(requestId))
            .toList()
            .doOnNext { addTripGroups(requestId, it) }
            .toBlocking()
            .first()
      }

  open fun getTripGroup(uuid: String): TripGroup? = tripGroupCache[uuid]

  open fun deletePastRoutesAsync(): Observable<Int> = routeStore.deleteAsync(
      WhereClauses.happenedBefore(
          3, /* hours */
          currentMillisProvider.call()
      ))

  open fun addTripGroups(requestId: String, groups: List<TripGroup>) {
    val tripGroupIds = groups.map { tripGroup ->
      addTripGroup(tripGroup)
      tripGroup.uuid()
    }
    if (requestIdTripGroupIdCache[requestId] == null) {
      requestIdTripGroupIdCache[requestId] = mutableListOf()
    }
    requestIdTripGroupIdCache[requestId]!!.addAll(tripGroupIds)
    onNewTripGroupsAvailable.onNext(requestId)
  }

  open fun addTripGroup(tripGroup: TripGroup) {
    tripGroupCache[tripGroup.uuid()] = tripGroup
  }

  open fun onNewTripGroupsAvailable(): Observable<String> = onNewTripGroupsAvailable.asObservable()

}